### ISSUE TITLE
How to resume LORA training ?

### ISSUE THREAD

**Issue by gitaiuser20** [Linaqruf/kohya-trainer#57]
i'd like to resume training the LORA, but i must be missing something because i'm not sure where to load the LORA and if there is any cell i need to tick to confirm that i'm resuming ?

**Comment by Linaqruf** [comment]
greetings
you can load your dataset from the latest training by downloading them and then put your LoRA weight to this `network_weights` field

![image](https://user-images.githubusercontent.com/50163983/217423473-c6ffa2ac-1d3d-463c-a4ec-753f5a93cafa.png)

then start training

**Comment by gitaiuser20** [comment]
do i load the path to the LORA safetensors file here ? thanks

**Comment by GeneralAwareness** [comment]
> do i load the path to the LORA safetensors file here ? thanks

I am wondering the same thing.

**Comment by Fannovel16** [comment]
> do i load the path to the LORA safetensors file here ? thanks

Yes, it works like Dreambooth training.

**Comment by Linaqruf** [comment]
> do i load the path to the LORA safetensors file here ? thanks

Yes

**Comment by Linaqruf** [comment]
There is another way to resume, but it's not efficient as you need to save 5gigs optimizer files with `--save_state` and load the folder with `--resume <path/to/optimizer>`

pros: you can keep reproducibility from latest training by doing that
cons: 5gb optimizer file size, and slower to learn as you literally continue the training

**Comment by matroxI** [comment]
> greetings you can load your dataset from the latest training by downloading them and then put your LoRA weight to this `network_weights` field

If I train 3 epochs from scratch, then continue to train for 3 more epochs, will the result be the same as if I continuously train 6 epochs from scratch?

**Comment by Linaqruf** [comment]
It will be different, because even though you're resuming training, you're essentially starting from scratch with the latest pre-trained LoRA model. This means that the reproducibility is broken.

However, I didn't say that if you train a model for 6 epochs from scratch and then train it again for another 6 epochs, the results will be the same. In theory, they won't be. The shuffling of captions during training can result in shuffled tags, leading to a different result for model 2 compared to model 1.

This means that even if you use the same architecture, hyperparameters, and training data, the final model that you obtain after training will differ each time you run the process, due to randomness introduced by weight initialization and data shuffling.

**Comment by younyokel** [comment]
Hey @Linaqruf I started my training at 20 epochs and 10 reps, which amounted to 9,600 steps. My Colab kicked me out of session, but it managed to save the 14th epoch in Google Drive (set to save every 2 epoch). When I import the Lora file with this setting, it starts training 9,600 steps from the first epoch. Is there a way of continuing from epoch 14 to better keep the track of things?

**Comment by Linaqruf** [comment]
> Hey @Linaqruf I started my training at 20 epochs and 10 reps, which amounted to 9,600 steps. My Colab kicked me out of session, but it managed to save the 14th epoch in Google Drive (set to save every 2 epoch). When I import the Lora file with this setting, it starts training 9,600 steps from the first epoch. Do I just make 6 more epochs then stop it or maybe is there a way of continuing from epoch 14 to better keep the track of things?

Hi, I'm sorry for my very late reply, because I'm struggling to simplify my notebook in this 2 months.

There is a way to resume training while also keep reproducibility and keep the track of the training, you need to set
1. `save_state` to `True`
2. Save the `last-state` things created by the config, file size around 5GB
3. Load that `last-state` folder in `resume_path` variable, or `--resume` hyperparameter

You can track your last training better than just specify LoRA weight to `network_weight`, but i'm not include that in my notebook because:
1. LoRA train faster than fine-tuning full model
2. saved `last-state` folder every n epoch need more memory, and it's not worth

Thanks

**Comment by Linaqruf** [comment]
I'm closing this issue because the main issue is already answered, but feel free to continue the discussion!

**Comment by orchidalloy** [comment]
It doesn't seem like --resume works. It says it loaded the previous save state, but it starts training from scratch, you can tell because the first epoch is shit even though it's supposedly continuing from the last one

**Comment by orchidalloy** [comment]
My seed is 42 in both cases. I think the feature is broken from upstream (anyone wanna call kohya?)

**Comment by Linaqruf** [comment]
I think there is no exact way to resuming from the last training. Currently, there are 3 method of resume training using Kohya's script:

1. Resume the training from last saved checkpoint and last saved optimizer state
2. Resume the training from pretrained model you use for training and last saved optimizer state
3. Resume the training straight from last saved checkpoint

This is for fine-tuning full model, I always go with option `2. Resume the training from pretrained model you use for training and last saved optimizer state` back then. LoRA training, you can change last saved checkpoint term with last saved LoRA aka `network_weight`. I never actually tried to resume LoRA training because it train faster and can do higher batch even in colab free tier. 

Using option 1 and 2 to continue the training gives you advantage to keep reproducibility from the last training. But it doesn't mean you actually continue from the last training. It just, you're using weight information from the last training as reference. While option 3 is literally `trained new LoRA from scratch` even though it using LoRA weight saved from the last training. You will get totally different result because no reproducibility is keep'ed.

This is an example for all 3 resuming method, I use WD 1.5 Beta 2 
1. `hitokomoru_locon_wd.safetensors`
    - Base LoRA weight 
    - num epochs = 10
    - Training seed = -1 (random)
    - Gen seed = -1 (random)
2. `hitokomoru_locon_wd_resume_state.safetensors`
    - Continued from hitokomoru_locon_wd.safetensors
    - Resume from last-state = True
    - Resume from network_weight = True
    - num epochs = 10
    - Training seed = 1337
    - Gen seed = 1337
3. `hitokomoru_locon_wd_resume_state_no_weigh.safetensors`
    - Continued from hitokomoru_locon_wd 
    - Resume from last-state = True
    - Resume from network_weight = False
    - num epochs = 10
    - Training seed = 1337
    - Gen seed = 1337
4. `hitokomoru_locon_wd_resume_state.safetensors`
    - Continued from hitokomoru_locon_wd 
    - Resume from last-state = False
    - Resume from network_weight = True
    - num epochs = 10
    - Training seed = 1337
    - Gen seed = 1337

Also, it actually resuming and not train from scratch, you can look all my prompt example in this repo, as well as config, oprimizer state, and LoRA weight

https://huggingface.co/Linaqruf/experimental/tree/main/hitokomoru_locon_wd/sample
 
![image](https://user-images.githubusercontent.com/50163983/227465718-bf5abc04-5f9f-4009-afed-da4abe0b2f7b.png)

**Comment by jianfuzhang** [comment]
Hello, thanks for the detailed explain! I'm using the Colab you provided. 
Didn't find any options that can set "save_state", and also, didn't figure out how to set `Resume from last-state` as well as `Resume from network_weight` option.
My understand is, if your loRA training dropped using the Colab provided. there is no way to resume your work. you can put your temp safetensors file in `network_weight` field, but it's the same thing with training from scratch.
if my understand is correct, why bother to put `network_weight` option there at all??
folks, Please let me know if my understanding is correct, very appreciated.

**Comment by Linaqruf** [comment]
Hi, I already said I'm not include and don't have a plan to include that (`save_state` and resume from `save_state`) in my notebook because:

1. LoRA train faster than fine-tuning full model
2. saved last-state folder every n epoch need more memory (approx 5GB), and it's not worth

But you can customize that by
1. Manually edit python dict in `5.4. Training Config` cell, you can find both `save_state` and/or `resume_state` at `training_arguments` section

```python
5.4. Training Config
...
    "training_arguments": {
        ....
        "save_state": <True/False>,
        "save_last_n_epochs_state": None,
        "resume": <path/to/optimizer/state>,
        ...
...
```
2. Manually edit `config_file.toml`, by adding `save_state` and/or `resume`
```toml

...
[training_arguments]
save_state = True/False
resume = "path/to/optimizer/state"
...

```
3. Manually edit training args in `5.5. Start Training`, by adding `--save_state and/or `--resume`
```python
5.5. Start Training

os.chdir(repo_dir)
!accelerate launch --config_file={accelerate_config} --num_cpu_threads_per_process=1 train_network.py \
  --sample_prompts={sample_prompt} \
  --dataset_config={dataset_config} \
  --config_file={config_file} \
  --save_state \
  --resume="path/to/optimizer/state"
```

**Comment by Linaqruf** [comment]
>As far as I understand, if the LoRA training process using the provided Colab is interrupted, there appears to be no way to resume the work. Even if I put the temp safetensors file in the network_weight field, it seems to be the same as starting the training process from scratch.

Hi, if you train LoRA for 10 epochs and you want to resume training using  `network_weight`  you need to set dim, alpha, and module/category the same as you train the first 10 epoch, or it will throw an error.

>I am wondering why the network_weight option is there in the first place if it cannot resume the training process. Could you please confirm whether my understanding is correct or not? Thank you very much for your help.

`network_weight` is used in training, you can see the syntax printed in the logs, it also saved to metadata. And you can see the comparison I provided earlier.

**Comment by Linaqruf** [comment]
Hi, there is a fix for `network_weights` are not working when training so I hope we can get better resuming method when this merged to main.

https://github.com/kohya-ss/sd-scripts/commit/83c7e03d050fc25f47a591c4ddfe28abdabc7ae7
